tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -288,6 +315,104 @@ bool SpdpHandler::do_normal_processing(const ACE_INET_Addr& a_remote,
   return true;
 }
 
+bool VerticalHandler::parse_message(OpenDDS::RTPS::MessageParser& message_parser,
+                                    ACE_Message_Block* msg,
+                                    bool& is_beacon,
+                                    bool check_submessages)
+{
+  if (!message_parser.parseHeader()) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: RelayHandler::parse_message failed to deserialize RTPS header\n"));
+    return false;
+  }
+
+  const auto& header = message_parser.header();
+  OpenDDS::DCPS::RepoId src_guid;
+  std::memcpy(src_guid.guidPrefix, header.guidPrefix, sizeof(OpenDDS::DCPS::GuidPrefix_t));
+  src_guid.entityId = OpenDDS::DCPS::ENTITYID_PARTICIPANT;
+
+  do {
+    if (!message_parser.parseSubmessageHeader()) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: RelayHandler::parse_message failed to deserialize RTPS submessage header\n"));
+      return false;
+    }
+
+    const auto submessage_header = message_parser.submessageHeader();
+
+    if (submessage_header.submessageId != OpenDDS::RTPS::PAD) {
+      is_beacon = false;
+    }
+
+    if (check_submessages && src_guid != application_participant_guid_) {
+#ifdef OPENDDS_SECURITY
+      switch (submessage_header.submessageId) {
+      case OpenDDS::RTPS::SRTPS_PREFIX:
+        {
+          if (application_participant_crypto_handle_ == DDS::HANDLE_NIL) {
+            ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: RelayHandler::parse_message no crypto handle for application participant\n"));
+            return false;
+          }
+
+          DDS::Security::ParticipantCryptoHandle remote_crypto_handle = rtps_discovery_->get_crypto_handle(application_domain_, application_participant_guid_, src_guid);
+          if (remote_crypto_handle == DDS::HANDLE_NIL) {
+            ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: RelayHandler::parse_message no crypto handle for client %C\n", guid_to_string(src_guid).c_str()));
+            return false;
+          }
+
+          DDS::OctetSeq encoded_buffer, plain_buffer;
+          DDS::Security::SecurityException ex;
+
+          encoded_buffer.length(msg->length());
```

Comment:
This doesn't support `ACE_Message_Block` chaining, which I think is fine for how the Relay is set up (AFAICT).  However it may be good to add an explicit to check for the `cont()` pointer being null.